File: 2021/10/13/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        self.i = 0
        n = len(preorder)
        
        def bfs(b):
            if n == self.i or preorder[self.i] > b:
                return None
            
            root = TreeNode(val=preorder[self.i])
            self.i += 1
            root.left = bfs(root.val)
            root.right = bfs(b)
            return root
        
        return bfs(1 << 32)

File: 2021/10/13/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bstFromPreorder(preorder []int) *TreeNode {
	i, n := 0, len(preorder)

	var bfs func(int) *TreeNode
	bfs = func(b int) *TreeNode {
		if i == n || preorder[i] > b {
			return nil
		}
		root := &TreeNode{
			Val: preorder[i],
		}
		i++
		root.Left = bfs(root.Val)
		root.Right = bfs(b)
		return root
	}

	return bfs(1 << 32)
}
